File: PyPoolBot/Que.py
import numpy as np
from Vector2 import Vector2
from NeuralNetwork import NeuralNetwork
from Table import Ball, Hole
import Utils
import Config
import math

class Que():

	def __init__(self, brain = None):
		self.angle = 0
		self.force = 0
		self.initialForce = 0

		self.x = 0
		self.y = 0

		self.ballsHit = False
		self.redPotted = False
		self.whitePotted = False
		self.totalScore = 0
		self.closestDistance = Config.WIDTH
		self.gamesPlayed = 0

		if brain != None:
			self.brain = brain.copy()
		else: 
			self.brain = NeuralNetwork(6, Config.HIDDEN_LAYERS, Config.HIDDEN_NODES, 2)
			# self.brain = NeuralNetwork(3, Config.HIDDEN_LAYERS, Config.HIDDEN_NODES, 2)

	@staticmethod
	def breed(brain1, brain2): 
		newbrain1, newbrain2 = NeuralNetwork.crossover(brain1, brain2)
		return Que(newbrain1), Que(newbrain2)

	def copy(self):
		c = Que(self.brain)
		c.ballsHit = self.ballsHit
		c.redPotted = self.redPotted
		c.whitePotted = self.whitePotted
		c.totalScore = self.totalScore
		c.gamesPlayed = self.gamesPlayed
		c.closestDistance = self.closestDistance
		return c

	def think(self, whiteBall, redBall, hole):
		wx = whiteBall.pos.x / Config.WIDTH
		wy = whiteBall.pos.y / Config.HEIGHT
		rx = redBall.pos.x / Config.WIDTH
		ry = redBall.pos.y / Config.HEIGHT
		hx = hole.pos.x / Config.WIDTH
		hy = hole.pos.y / Config.HEIGHT
		inputs = np.array([wx, wy, rx, ry, hx, hy])
		# inputs = np.array([wy, ry, hy])
		# print(inputs)
		output = self.brain.predict(inputs)
		# print(output)
		self.angle = Utils.map(output[0], 0, 1, 0, math.pi, False)
		self.force = Utils.map(output[1], 0, 1, 0.5, 1, False)
		self.initialForce = self.force

	def mutate(self):
		self.brain.mutate(0.1)

	def move_to(self, x, y):
		self.x = x
		self.y = y

	def get_force(self):
		w = Vector2.fromAngle(self.angle - math.pi / 2, 1)
		w.setMag(self.initialForce * 10)
		return w

	def update_score(self, hole, ball):
		dist = ball.pos.dist(hole.pos)
		if dist < self.closestDistance:
			self.closestDistance = dist

	def apply_score(self):
		score = 0
		if self.redPotted:
			score = 1
		elif self.whitePotted:
			score = 0
		elif self.ballsHit == False:
			score = 0
		else:
			score = (1 - (self.closestDistance / Config.WIDTH)) / 2

		self.totalScore += score;
		self.gamesPlayed += 1

		self.redPotted = False
		self.whitePotted = False
		self.ballsHit = False
		self.closestDistance = Config.WIDTH

		return self.totalScore / self.gamesPlayed

	def draw(self, graphics):
		v = Vector2(self.x, self.y, 0)
		w = Vector2.fromAngle(self.angle - math.pi / 2, 1)
		w.setMag(15 + (20 if self.force > 0 else 0))
		p1 = Vector2.subv(v, w)
		w.setMag(100)
		p2 = Vector2.subv(v, w)

		graphics.line(p1.x, p1.y, p2.x, p2.y, 8, (150, 75, 0))

File: PoolBot/que.js
// @ts-check
"use strict";

var Vector2 = require('./vector2.js');
var NeuralNetwork = require('./nn.js');

class Que {
    constructor(brain) {

        /*let input1 = random();
        let input2 = random();

        this.angle = map(input1, 0, 1, 0, PI);
        this.force = map(input2, 0, 1, 0.5, 1);*/
        this.angle = 0;
        this.force = 0;
        this.initialForce = 0;

        this.ballsHit = false;
        this.redPotted = false;
        this.whitePotted = false;
        this.totalScore = 0;
        this.closestDistance = global.width;
        this.gamesPlayed = 0;

        if (brain) {
            this.brain = brain.copy();
        } else {
            this.brain = new NeuralNetwork(6, 18, 2);
        }
    }

    copy() {
        var c = new Que(this.brain);
        c.ballsHit = this.ballsHit;
        c.redPotted = this.redPotted;
        c.whitePotted = this.whitePotted;
        c.totalScore = this.totalScore;
        c.gamesPlayed = this.gamesPlayed;
        c.closestDistance = this.closestDistance;
        return c;
    }

    think(whiteBall, redBall, hole) {
        let inputs = [];
        inputs[0] = whiteBall.pos.x / global.width;
        inputs[1] = whiteBall.pos.y / global.height;
        inputs[2] = redBall.pos.x / global.width;
        inputs[3] = redBall.pos.y / global.height;
        inputs[4] = hole.pos.x / global.width;
        inputs[5] = hole.pos.y / global.height;
        let output = this.brain.predict(inputs);
        this.angle = global.map(output[0], 0, 1, 0, Math.PI);
        this.force = global.map(output[1], 0, 1, 0.5, 1);
        this.initialForce = this.force;
    }

    mutate() {
        this.brain.mutate(0.10);
    }

    moveTo(x, y) {
        this.x = x;
        this.y = y;
    }

    getForce() {
        let w = Vector2.fromAngle(this.angle - Math.PI / 2);
        w.setMag(this.initialForce * 10);
        return w;
    }

    updateScore(hole, ball) {
        var dist = ball.pos.dist(hole.pos);
        if (dist < this.closestDistance) {
            this.closestDistance = dist;
        }
    }

    applyScore() {
        let score = 0;
        if (this.redPotted) {
            score = 1
        }
        else if (this.whitePotted) {
            score = 0;
        }
        else if (this.ballsHit === false) {
            score = 0;
        }
        else {
            score = (1 - (this.closestDistance / global.width)) ;
        }

        this.totalScore += score;
        this.gamesPlayed++;

        this.redPotted = false;
        this.whitePotted = false;
        this.ballsHit = false;
        this.closestDistance = global.width;

        return this.totalScore / this.gamesPlayed;
    }

    draw() {
        global.strokeWeight(8);
        global.stroke(150, 75, 0);

        let v = new Vector2(this.x, this.y);
        let w = Vector2.fromAngle(this.angle - Math.PI / 2);
        w.setMag(15 + (this.force > 0 ? 20 : 0));
        let p1 = Vector2.sub(v, w);
        w.setMag(100);
        let p2 = Vector2.sub(v, w);

        global.line(p1.x, p1.y, p2.x, p2.y);

        global.strokeWeight(1);
    }
}

if (typeof module !== 'undefined') {
    module.exports = Que;
}
